### PR TITLE
Causal attention

### PR DESCRIPTION
**Pull request by cifkao**
Implementing causal ("unidirectional") attention. Seems to work:

```pycon
>>> att = FAVOR(16, redraw=False, multihead=False)
>>> k = torch.randn(1, 16, 10)
>>> q = torch.randn(1, 16, 10)
>>> v = torch.arange(1., 21.).view(1, 2, 10)
>>> v
tensor([[[ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.],
         [11., 12., 13., 14., 15., 16., 17., 18., 19., 20.]]])
>>> att(k, v, q)
tensor([[[ 5.4014,  5.4832,  5.3723,  5.6341,  5.9119,  5.5918,  5.7551,
           5.4221,  5.2524,  5.4419],
         [15.4014, 15.4832, 15.3723, 15.6341, 15.9119, 15.5918, 15.7551,
          15.4221, 15.2524, 15.4419]]])
>>> att.causal = True
>>> att(k, v, q)
tensor([[[ 1.0000,  1.5649,  1.9576,  2.4059,  2.8802,  3.1429,  3.5695,
           4.6532,  4.7577,  5.4419],
         [11.0000, 11.5649, 11.9576, 12.4059, 12.8802, 13.1429, 13.5695,
          14.6532, 14.7577, 15.4419]]])
```

Also implementing multi-head attention with `multihead=True`.

### CODE REVIEW

File: favor/favor.py [hunk 1]
```diff
@@ -87,22 +97,39 @@ def phi(x):
...
+            out = torch.matmul(         # (batch, n, dim)
+                phi_q[:, :, None, :],   # (batch, n, 1, r)
+                k_v_prod.cumsum(dim=1)  # (batch, n, r, dim)
+            ).squeeze(2)
+
+            # normalization factors: (batch, n, 1)
+            norm = torch.matmul(
+                phi_q[:, :, None, :],           # (batch, n, 1, r)
+                phi_k.cumsum(dim=1)[..., None]  # (batch, n, r, 1)
+            ).squeeze(2)
```
Comment (aliutkus) [inline favor/favor.py]: are you 100% sure ?

File: favor/favor.py [hunk 2]
```diff
@@ -15,6 +13,8 @@ def __init__(
         self,
         key_dim,
         orthonormal=True,
+        multihead=True,
```
Comment (aliutkus) [inline favor/favor.py]: I don't really like including the multihead in the base class, because it makes the thing less readable IMO

could you create a derived MultiHeadFAVOR, that create several FAVOR ? This would not be equivalent ?

Comment (cifkao) [reply favor/favor.py]: This is just for convenience, equivalently you can just reshape before and after the attention. So yes, maybe we can instead have a wrapper which will do that.

Another option would be to just allow any number of dimensions before *and* after the actual feature dimension (and just flatten them as you already do). So for single-head attention we would have `dim=1`, for multi-head `dim=2`; if for some reason you didn't have batch dimension, you could have `dim=0`...

File: favor/favor.py (hunk 1 above)
Comment (cifkao) [reply favor/favor.py]: Hmm, maybe not 100%. :smile: We could write a test to check that we can approximate standard attention with a causal attention mask (with the softmax kernel).

Comment (aliutkus) [reply favor/favor.py]: ok, what about memory usage ? we never store the big attention, ok ?

File: favor/favor.py (hunk 2 above)
Comment (aliutkus) [reply favor/favor.py]: thing is: IMO, multihead attention is indeed just a matter ok the transforms to obtain the keys, values, queries, right ? The actual attention layer is the same in all cases. I think that flattening is indeed the way to go

File: favor/favor.py (hunk 1 above)
Comment (cifkao) [reply favor/favor.py]: Yes, that's OK, the `n` dimension is always one of the non-matrix ("batch") dimensions in the `matmul`. It's the trick from Appendix B.1.

File: favor/favor.py (hunk 2 above)
Comment (cifkao) [reply favor/favor.py]: OK, reverting the multi-head thing.

File: favor/favor.py [hunk 3]
```diff
@@ -51,20 +51,21 @@ def features(self):
...
+        values: (batch, values_dimension, *keys_locations)
         queries: (batch, keys_dimension, *queries_locations)
         """
         # flattening everything
         keys_locations = keys.shape[2:]
         queries_locations = queries.shape[2:]
-        keys = keys.view(*keys.shape[:2], -1)
-        values = values.view(*values.shape[:2], -1)
-        queries = queries.view(*queries.shape[:2], -1)
+        keys, values, queries = (x.view(*x.shape[:2], -1) for x in (keys, values, queries))
```
Comment (aliutkus) [inline favor/favor.py]: ok, but please PEP9

File: favor/favor.py [hunk 4]
```diff
@@ -51,20 +51,21 @@ def features(self):
...
+        if self.causal and keys_locations != queries_locations:
+            raise ValueError(
+                'Expected equal key and query locations with causal attention, got: '
+                '{}, {}'.format(keys_locations, queries_locations))
 
         # getting to (batch, n, dim)
-        keys = keys.permute(0, 2, 1)
-        values = values.permute(0, 2, 1)
-        queries = queries.permute(0, 2, 1)
+        keys, values, queries = (x.permute(0, 2, 1) for x in (keys, values, queries))
```
Comment (aliutkus) [inline favor/favor.py]: ok but please PEP9 (line length)

File: favor/favor.py (hunk 1 above)
Comment (aliutkus) [reply favor/favor.py]: all right, awesome

File: favor/favor.py [hunk 5]
```diff
@@ -87,20 +88,35 @@ def phi(x):
...
-            torch.sum(phi_k, dim=1)[..., None] # (batch, r, 1)
-        ) 
+            # normalization factors: (batch, n, 1)
+            norm = torch.matmul(
+                phi_q,
+                phi_k.sum(dim=1)[..., None] # (batch, r, 1)
+            )
 
-        out = out / scale
+        out /= norm
```
Comment (aliutkus) [inline favor/favor.py]: no, be careful that this probably does not propagate gradient: we should not do in place operations. or has this been fixed in latest pytorch ?

**Review by aliutkus — COMMENTED** [review]
just some PEP9 for line length, this thing about inplace assignment at the end, and good to go to me

File: favor/favor.py (hunk 5 above)
Comment (cifkao) [reply favor/favor.py]: I remember reading [this](https://pytorch.org/docs/stable/notes/autograd.html#in-place-correctness-checks):
> [...] if you’re using in-place functions and not seeing any errors, you can be sure that the computed gradients are correct.

So that means it should be safe, no? (unless there's a bug I don't know about)

Comment (aliutkus) [reply favor/favor.py]: please revert just this and then I'll merge
I really encountered errors due to that in the past.

_or_: really test backprop through this, not only forward

Comment (cifkao) [reply favor/favor.py]: OK!

**Comment by aliutkus** [comment]
perfect, thanks
